app/assets/javascripts/discourse/app/routes/topic.js

```diff
@@ -327,6 +328,12 @@ const TopicRoute = DiscourseRoute.extend({
   activate() {
     this._super(...arguments);
     this.set("isTransitioning", false);
+
+    const topic = this.modelFor("topic");
+    this.session.set("lastTopicIdViewed", {
+      topicId: parseInt(topic.get("id"), 10),
+      historyUuid: this.router.location.getState?.().uuid,
+    });
```

Comment:
This is where the logic used to live before https://github.com/discourse/discourse/commit/30c152c5a7127790a16482321b68650514ad16c9. I moved it out of here for two reasons:

1. We want to store the focus information against the uuid of the topic list route, not the topic route. I think the reason it actually works here is because, when `activate` is triggered, Ember hasn't yet called the browser's `history.pushState`. Relying on precise ordering of Ember's router feels a bit risky... 

   That said, we have a robust test so should detect any Ember changes in future. So I could live with it... but...

2. It means that **any** method of navigating to a topic will cause the ID to be stored. This is problematic because there are many other ways to navigate from a topic list route to a topic route. For example, it could happen via a notification, a custom sidebar link, or a link introduced by a theme.

   In those cases, I don't think we want topic-list focus to be set to the topic you visited via one of those other methods.

   It would be good to add a test for at least one of those, so we don't regress in future.

To avoid both these issues, I think this logic belongs in the topic-list-item component. I suppose we'll need to find some way of detecting when the 'last activity' link is followed 🤔

Reply:
Yeah, that makes sense, thanks @davidtaylorhq! I have pushed a fix that

- switches back to storing the focus information in the `topic-list-item` route
- adds an event handler when using `Enter` on the `activity` element
- adds handling for the Replies column's go to first post/go to last post buttons

In writing this, I am realizing I am missing one additional use case: clicking on the Activity column link still doesn't trigger the calculation, will add that as well. 

(This is the tradeoff with doing this in `topic-list-item` though, we need to account for each use case individually.)